### ISSUE TITLE
 [dpc++] reduce_by_segment + make_device_policy mangled name redefinition error

### ISSUE THREAD

**Issue by abagusetty** [oneapi-src/oneDPL#507]
Specifically related to using `reduce_by_segment` (non-range version) when using the function multiple times in the same translation unit without passing a unique lambda name for `make_device_policy`.

Compiler: Intel(R) oneAPI DPC++/C++ Compiler 2022.1.0 (2022.x.0.20220112)
Bears some similarity with #166 

```
../include/sycl/CL/sycl/handler.hpp:203:7: error: definition with same mangled name '_ZTSN2cl4sycl6detail19__pf_kernel_wrapperIN6oneapi3dpl20__par_backend_hetero20__unique_kernel_nameINS4_9execution5__dpl13device_policyINS8_17DefaultKernelNameEEELm3EEEEE' as another definition
class RoundedRangeKernel {
      ^
/soft/restricted/CNDA/sdk/2022.01.30.001/oneapi/compiler/pseudo-20220112/compiler/linux/bin-llvm/../include/sycl/CL/sycl/handler.hpp:203:7: note: previous definition is here
In file included from IJVector_parcsr_device.c:15:
In file included from ./../utilities/_hypre_utilities.hpp:1100:
In file included from /home/abagusetty/build/oneDPL/build_sycl/include/oneapi/dpl/execution:61:
In file included from /home/abagusetty/build/oneDPL/build_sycl/include/oneapi/dpl/pstl/algorithm_impl.h:31:
In file included from /home/abagusetty/build/oneDPL/build_sycl/include/oneapi/dpl/internal/../pstl/parallel_backend.h:19:
/home/abagusetty/build/oneDPL/build_sycl/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:244:75: error: definition with same mangled name '_ZTSN6oneapi3dpl20__par_backend_hetero20__unique_kernel_nameINS0_9execution5__dpl13device_policyINS4_17DefaultKernelNameEEELm3EEE' as another definition
            __cgh.parallel_for<_Name...>(sycl::range</*dim=*/1>(__count), [=](sycl::item</*dim=*/1> __item_id) {
                                                                          ^
/home/abagusetty/build/oneDPL/build_sycl/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:244:75: note: previous definition is here
/home/abagusetty/build/oneDPL/build_sycl/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:470:79: error: definition with same mangled name '_ZTSN6oneapi3dpl20__par_backend_hetero6__pslkIJNS1_20__unique_kernel_nameINS0_9execution5__dpl13device_policyINS5_17DefaultKernelNameEEELm1EEEEEE' as another definition
                sycl::nd_range<1>(__n_groups * __wgroup_size, __wgroup_size), [=](sycl::nd_item<1> __item) {
                                                                              ^
/home/abagusetty/build/oneDPL/build_sycl/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:470:79: note: previous definition is here
/home/abagusetty/build/oneDPL/build_sycl/include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl.h:493:70: error: definition with same mangled name '_ZTSN6oneapi3dpl20__par_backend_hetero7__psgrkIJNS1_20__unique_kernel_nameINS0_9execution5__dpl13device_policyINS5_17DefaultKernelNameEEELm1EEEEEE' as another definition
                    sycl::nd_range<1>(__wgroup_size, __wgroup_size), [=](sycl::nd_item<1> __item) {
                                                                     ^
```

**Comment by dmitriy-sobolev** [comment]
@abagusetty, I was able to reproduce the issue using this code:

```c++
#include <oneapi/dpl/execution>
#include <oneapi/dpl/algorithm>

int main()
{
    constexpr std::size_t size = 10;
    int in_keys[size] = {0, 0, 0, 1, 1, 1, 2, 2, 2, 3};
    int in_vals[size] = {1, 1, 1, 1, 1, 1, 1, 1, 1, 1};

    int out_keys[size] = {-1 , -1, -1, -1, -1, -1, -1, -1, -1, -1};
    int out_vals[size] = {-1 , -1, -1, -1, -1, -1, -1, -1, -1, -1};

    int out_keys2[size] = {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1};
    double out_vals2[size] = {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1};

    sycl::queue queue{};
    auto policy = oneapi::dpl::execution::make_device_policy(queue);

    oneapi::dpl::reduce_by_segment(policy, in_keys, in_keys + size, in_vals, out_keys, out_vals);
    oneapi::dpl::reduce_by_segment(policy, in_keys, in_keys + size, in_vals, out_keys, out_vals2);

    return 0;
}
```

Is it easy for you to use a policy with a unique kernel name when calling algorithms? For example:
```c++
#include <oneapi/dpl/execution>
#include <oneapi/dpl/algorithm>

int main()
{
    constexpr std::size_t size = 10;
    int in_keys[size] = {0, 0, 0, 1, 1, 1, 2, 2, 2, 3};
    int in_vals[size] = {1, 1, 1, 1, 1, 1, 1, 1, 1, 1};

    int out_keys[size] = {-1 , -1, -1, -1, -1, -1, -1, -1, -1, -1};
    int out_vals[size] = {-1 , -1, -1, -1, -1, -1, -1, -1, -1, -1};

    int out_keys2[size] = {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1};
    double out_vals2[size] = {-1, -1, -1, -1, -1, -1, -1, -1, -1, -1};

    sycl::queue queue{};
    auto policy1 = oneapi::dpl::execution::make_device_policy<class reduce_by_segment1>(queue);
    auto policy2 = oneapi::dpl::execution::make_device_policy<class reduce_by_segment2>(policy1);

    oneapi::dpl::reduce_by_segment(policy1, in_keys, in_keys + size, in_vals, out_keys, out_vals);
    oneapi::dpl::reduce_by_segment(policy2, in_keys, in_keys + size, in_vals, out_keys, out_vals2);

    return 0;
}
```

Meanwhile, I am investigating the issue.

**Comment by dmitriy-sobolev** [comment]
@abagusetty, sorry for a long time to respond. 

Many algorithms are prone to that issue when used with Range-based API:
- for_each
- copy
- swap_ranges
- reverse
- reverse_copy
- rotate_copy
- transform
- replace_if
- replace
- replace_copy_if
- replace_copy
- move
- merge
- reduce_by_segment

reduce_by_segment is an exception. It is the only non-ranged based API prone to the issue. It is related to the fact that the non-ranged version shares implementation with the ranged version.

**Comment by abagusetty** [comment]
@dmitriy-sobolev No worries, thanks for the info. This really helps with planning

**Comment by dmitriy-sobolev** [comment]
@abagusetty, the error has been fixed by #521. Feel free to reopen the issue if you face it again.
